core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java

```diff
@@ -54,14 +57,26 @@ public long creationTime(TimeUnit unit) {
 
   public void setExpirationTime(long expirationTime, TimeUnit unit) {
     if (expirationTime == NO_EXPIRE) {
-      this.expirationTime = NO_EXPIRE;
+      updateExpirationTime(NO_EXPIRE);
     } else if (expirationTime <= 0) {
       throw new IllegalArgumentException("invalid expiration time: " + expirationTime);
     } else {
-      this.expirationTime = nativeTimeUnit().convert(expirationTime, unit);;
+      updateExpirationTime(nativeTimeUnit().convert(expirationTime, unit));
     }
   }
 
+  private void updateExpirationTime(long update) {
+    do {
+      long current = this.expirationTime;
+      if (current >= update) {
+        break;
+      }
+      if (EXPIRATIONTIME_UPDATER.compareAndSet(this, current, update)) {
+        break;
+      }
+    } while (true);
```

Comment:
Given the condition, I am in favour of a `while(true)` loop that has the benefit of clearly indicating you are doing something special.